[knx] Improve localization

holgerfriedrich opened the pull request:
- introduce localization of error messages
    - add new strings for common exceptions
    - provide helper functions for translation
    - add test cases

Signed-off-by: Holger Friedrich <mail@holger-friedrich.de>

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").

ATTENTION: Don't use "git merge" when working with your pull request branch!
This can clutter your Git history and make your PR unuseable.
Use "git rebase" instead. See this forum post for further details:
https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

Add one or more appropriate labels to make your PR show up in the release notes.
E.g. enhancement, bug, documentation, new binding
This can only be done by yourself if you already contributed to this repo.

If your PR's code is not backward compatible with previous releases (which
should be avoided), add a message to the release notes by filing another PR:
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

# Title

Provide a short summary in the *Title* above. It will show up in the release notes.
For example:
- [homematic] Improve communication with weak signal devices
- [timemachine][WIP] Initial contribution

# Description

Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work

# Testing

Your pull request will automatically be built and available under the following folder:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

It is a good practice to add a URL to your built JAR in this pull request description,
so it is easier for the community to test your Add-on.
If your pull request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Marketplace:
https://community.openhab.org/c/marketplace/69

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

holgerfriedrich commented:
This is still work in progress. It adds German translations to the KNX binding.

I would appreciate any feedback on the general approach how to add the translations.

1. property files: German translations are based on latest Crowdin export. Is it a valid approach to copy the export to get a start for a new language?
2. property files: I have  added a few lines which are not part of the offical binding yet. If I got it correctly, I should get this into knx.propeties. Once it is merged on master, it will show up on Crowdin and can be translated. When will the automatic PR process start for a language? What is the trigger for the automatic PR process? Who approves the changes on Crowdin?
3. property files: Will the automatic PR process wipe out all unknown lines?
4. property files: Any hints regarding namespace for additional strings/error messages?
5. LocaleProvider.getlocale() -> will it always return the same locale or does this change dynamically?

Thanks!

holgerfriedrich commented:
@lolodomo could you give some input to the questions above or recommend whom to ask? Thanks!

jlaur commented:
> I would appreciate any feedback on the general approach how to add the translations.

This page might also shed some light: https://www.openhab.org/docs/developer/utils/i18n.html#managing-translations

> 1. property files: German translations are based on latest Crowdin export. Is it a valid approach to copy the export to get a start for a new language?

I don't understand this question, can you elaborate?

> 2. property files: I have  added a few lines which are not part of the offical binding yet. If I got it correctly, I should get this into knx.propeties. Once it is merged on master, it will show up on Crowdin and can be translated.

Yes.

> When will the automatic PR process start for a language?

At least once per day.

> What is the trigger for the automatic PR process?

I believe it's a time trigger, but it can also be triggered manually by admins.

> Who approves the changes on Crowdin?

Preferably proof-readers, but I'm not sure we have many yet. So actually, this question I cannot answer - @lolodomo, can you?

> 3. property files: Will the automatic PR process wipe out all unknown lines?

Yes.

> 4. property files: Any hints regarding namespace for additional strings/error messages?

Each binding has its own namespace, so you don't need to use any prefixes to secure uniqueness. Best advice for custom texts would be to try to avoid potential conflicts with framework-generated strings/prefixes.

> 5. LocaleProvider.getlocale() -> will it always return the same locale or does this change dynamically?

It will change dynamically when user changes regional settings.

holgerfriedrich commented:
@jlaur thanks for the feedback.
I think I will add a test and then set this PR to ready for review the next days.

> > I would appreciate any feedback on the general approach how to add the translations.

> > 1. property files: German translations are based on latest Crowdin export. Is it a valid approach to copy the export to get a start for a new language?
> 
> I don't understand this question, can you elaborate?
>

This was about how to introduce a new language to a binding. Is it intended to download the property file for the new language from Crowdin and add it to the binding - and after this the bot takes over for further updates? Or should I just add empty files like ``knx_de.properties`` and the bot will take over as well?
 

> > Who approves the changes on Crowdin?
> 
> Preferably proof-readers, but I'm not sure we have many yet. So actually, this question I cannot answer - @lolodomo, can you?
> 

That would be interesting. We seem to have unapproved translations for KNX (de). According to the docs the bot will not include them, correct?

jlaur commented:
> > > 1. property files: German translations are based on latest Crowdin export. Is it a valid approach to copy the export to get a start for a new language?
> > 
> > 
> > I don't understand this question, can you elaborate?
> 
> This was about how to introduce a new language to a binding. Is it intended to download the property file for the new language from Crowdin and add it to the binding - and after this the bot takes over for further updates? Or should I just add empty files like `knx_de.properties` and the bot will take over as well?

Neither. New languages are automatically provided from Crowdin.

> > > Who approves the changes on Crowdin?
> > 
> > Preferably proof-readers, but I'm not sure we have many yet. So actually, this question I cannot answer - @lolodomo, can you?
> 
> That would be interesting. We seem to have unapproved translations for KNX (de). According to the docs the bot will not include them, correct?

Correct.

holgerfriedrich commented:
> > This was about how to introduce a new language to a binding. Is it intended to download the property file for the new language from Crowdin and add it to the binding - and after this the bot takes over for further updates? Or should I just add empty files like `knx_de.properties` and the bot will take over as well?
> 
> Neither. New languages are automatically provided from Crowdin.

@jlaur For me it is still not clear what actually triggers a new language file to be added in the first place. For example, there is a ``knx.properties``, but no file ``knx_de.properties`` in ``resources/OH-INF/i18n`` yet. Though there are translations available at Crowdin, the bot does not create a PR. I suppose this is because ``knx_de.properties`` does not exist.

jlaur commented:
> For me it is still not clear what actually triggers a new language file to be added in the first place. For example, there is a `knx.properties`, but no file `knx_de.properties` in `resources/OH-INF/i18n` yet. Though there are translations available at Crowdin, the bot does not create a PR. I suppose this is because `knx_de.properties` does not exist.

The important sentence from the documentation is this one: "Please note that all texts must be translated and approved in order to be included." So for a language to be included in PR's created from Crowdin, 100% of strings must be translated, and 100% must be approved. So it is not because `knx_de.properties` does not exist, but because at least one of those criteria are not fulfilled.

In German only 82% are approved. @cweitkamp, would you be able to have a look at the pending knx translations and/or possible shed some more light on the approval process?

cweitkamp commented:
> In German only 82% are approved. @cweitkamp, would you be able to have a look at the pending knx translations and/or possible shed some more light on the approval process?

Done.

jlaur commented:
@holgerfriedrich - the German translation is now merged: #13448.

holgerfriedrich commented:
thanks, @cweitkamp !
@jlaur I added test cases and removed ``knx_de.properties`` from the commit as it is now already merged. I will introduce new strings via CrowdIn once this PR is merged. I took over you mock classes from hdpowerview.

holgerfriedrich commented:
@kaikreuzer could you please have a look? I'd Like to close this before getting deeper into the data secure topic again... Thanks!

jlaur commented on bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/i18n/KNXTranslationProviderTest.java:
```diff
@@ -0,0 +1,118 @@
+        assertEquals(UNKNOWN, KNXTranslationProvider.I18N.get(UNKNOWN, 5));
+        assertEquals(UNKNOWN_NULL, KNXTranslationProvider.I18N.get(UNKNOWN_PATTERN, null, null));
+        assertEquals(UNKNOWN_FIVE, KNXTranslationProvider.I18N.get(UNKNOWN_PATTERN, 5));
+        // this would cause a compiler warning as it cannot be casted to Object[]
+        // assertNotNull(KNXTranslationProvider.I18N.get(UNKNOWN, null));
+        // assertNotNull(KNXTranslationProvider.I18N.get(UNKNOWN_PATTERN, null));
```
Can this be omitted?

holgerfriedrich replied:
I activated this test and added a SuppressWarnings to the test case.... Maybe better to check this edge cases as well.

jlaur commented on bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/i18n/KNXTranslationProviderTest.java:
```diff
@@ -0,0 +1,118 @@
+        final Exception se = new KNXLinkClosedException("connection closed", e);
+        assertNotNull(KNXTranslationProvider.I18N.getLocalizedException(e));
+        assertNotNull(KNXTranslationProvider.I18N.getLocalizedException(se));
+        // assertEquals("Exception", KNXTranslationProvider.I18N.getLocalizedException(e));
```
Can this be omitted?

jlaur commented on bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/i18n/MockedTranslationProvider.java:
```diff
@@ -0,0 +1,68 @@
+public class MockedTranslationProvider implements TranslationProvider {
+
+    private static final Map<String, String> TEXTS = Map.ofEntries(
+            entry("dynamic-channel.scene-activate.description", "Activates the scene ''{0}''"),
+            entry("dynamic-channel.scene-group-activate.description", "Activates the scene group ''{0}''"),
+            entry("dynamic-channel.automation-enabled.description", "Enables/disables the automation ''{0}''"),
+            entry("dynamic-channel.automation-enabled.label", "{0}, {1}, {2}"),
+            entry("dynamic-channel.automation.hour", "{0}hr"), entry("dynamic-channel.automation.minute", "{0}m"),
+            entry("dynamic-channel.automation.hour-minute", "{0}hr {1}m"),
+            entry("dynamic-channel.automation.at-sunrise", "At sunrise"),
+            entry("dynamic-channel.automation.before-sunrise", "{0} before sunrise"),
+            entry("dynamic-channel.automation.after-sunrise", "{0} after sunrise"),
+            entry("dynamic-channel.automation.at-sunset", "At sunset"),
+            entry("dynamic-channel.automation.before-sunset", "{0} before sunset"),
+            entry("dynamic-channel.automation.after-sunset", "{0} after sunset"),
+            entry("dynamic-channel.automation.weekdays", "Weekdays"),
+            entry("dynamic-channel.automation.weekends", "Weekends"),
+            entry("dynamic-channel.automation.all-days", "All days"),
```
These are copied from hdpowerview and are probably useless here.

jlaur commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/i18n/KNXTranslationProvider.java:
```diff
@@ -0,0 +1,130 @@
+ *
+ */
+@NonNullByDefault
+public enum KNXTranslationProvider {
```
Why is this implemented as **enum**? You could consider implementing it as a **class** and inject `LocaleProvider` and `TranslationProvider` in constructor. This way you would also avoid some nullness issues. See for example:
https://github.com/openhab/openhab-addons/blob/6e4f78458d143ced35485241ab1ec87c0990dcc9/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/MieleTranslationProvider.java#L29-L60

holgerfriedrich replied:
I wanted to implement the TranslationProvider as a static singleton. This makes it easy to use from any class within the binding. Injecting LocaleProvider and TranslationProvider is done in KNXHandlerFactory, which hands it over to the static TranslationProvider instance.

I tried implement it differently before by propagating LocaleProvider and TranslationProvider all the way down from KNXHandlerFactory to all classes which require the translation service.... This was quite a mess and the static singleton approach looks much clearer.

>  * @implNote It is implemented as a static singelton, enforced by the single-element enum pattern.

Implementing a singleton using an enum is actually a quite nice pattern from the boot "Effective Java". It may look a bit strange when reading the code the first time :-)

jlaur replied:
OK.

jlaur commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/i18n/KNXTranslationProvider.java:
```diff
@@ -0,0 +1,130 @@
+     * @return localized message in form <description (translated)> (<class name>, <e.getLocalizedMessage (not
+     *         translated)>), empty string for null. May possibly change in further releases.
+     */
+    public String getLocalizedException(final @Nullable Throwable e) {
+        if (e == null) {
+            return "";
+        }
```
This seems like potentially hiding a bug, consider making parameter `@NonNull`.

jlaur commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/i18n/KNXTranslationProvider.java:
```diff
@@ -0,0 +1,130 @@
+        final TranslationProvider translationProvider = this.translationProvider;
+        final LocaleProvider localeProvider = this.localeProvider;
+        if (translationProvider != null) {
+            // localeProvider might be null, but if not, getLoacle will return NonNull Locale
```
```suggestion
            // localeProvider might be null, but if not, getLocale will return NonNull Locale
```

jlaur reviewed (changes requested):
Thanks! I've added some feedback. Perhaps @kaikreuzer would like to have a second look.

holgerfriedrich commented:
@jlaur thanks for the review! I think I addressed all of your comments. I left one conversation open, explaining why I used the enum to implement the singleton.

kaikreuzer commented:
> @kaikreuzer could you please have a look?

@jlaur is the much better reviewer for this, so he can merge right away once he is done.

jlaur commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -87,11 +88,11 @@ public void initializeLater() {
-            }
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "KNX security: " + message);
+            Throwable cause = e.getCause();
+            if (cause == null)
+                cause = e;
```
```suggestion
            if (cause == null)
            {
                cause = e;
            }
```

jlaur commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/i18n/KNXTranslationProvider.java:
```diff
@@ -0,0 +1,127 @@
+            // locale cannot be cached, as getLocale() will return different result once locale is changed by user
+            final Locale locale = (localeProvider != null) ? localeProvider.getLocale() : Locale.getDefault();
+            final String res = translationProvider.getText(bundle, text, text, locale, arguments);
+            if (res != null)
+                return res;
```
```suggestion
            if (res != null)
            {
                return res;
            }
```

jlaur commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/i18n/KNXTranslationProvider.java:
```diff
@@ -0,0 +1,127 @@
+                return res;
+        }
+        // translating not possible, we still have the original text without any subsititutions
+        if ((arguments == null) || (arguments.length == 0)) {
```
Redundant parenthesis:
```suggestion
        if (arguments == null || arguments.length == 0) {
```

jlaur commented on bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/i18n/KNXTranslationProviderTest.java:
```diff
@@ -0,0 +1,115 @@
+
+    @Test
+    public void testGetBeforeInit() {
+        // NonNull, compilation error
+        // assertNull(KNXTranslationProvider.I18N.get(null));
```
Can this be removed?

jlaur commented on bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/i18n/KNXTranslationProviderTest.java:
```diff
@@ -0,0 +1,115 @@
+    static final String UNKNOWN_PATTERN = "unknown text {0}";
+    static final String UNKNOWN_FIVE = "unknown text 5";
+    static final String UNKNOWN_NULL = "unknown text null";
+    static final String KNOWN_DAY_KEY = "dynamic-channel.automation.weekdays";
+    static final String KNOWN_DAY_VALUE = "Weekdays";
+    static final String KNOWN_DAY_PATTERN_KEY = "dynamic-channel.automation.after-sunset";
+    static final String KNOWN_DAY_PATTERN_VALUE = "{0} after sunset";
+    static final String KNOWN_DAY_FIVE_VALUE = "5 after sunset";
+    static final String KNOWN_DAY_ORIG_VALUE = "null after sunset";
+    static final String KNOWN_DAY_NULL_VALUE = "null after sunset";
```
I'm wondering if it would make sense with some strings from the knx binding here? The mentioned keys do not exist in as properties, so it seems quite artificial.

jlaur commented on bundles/org.openhab.binding.knx/src/test/java/org/openhab/binding/knx/internal/i18n/MockedTranslationProvider.java:
```diff
@@ -0,0 +1,55 @@
+public class MockedTranslationProvider implements TranslationProvider {
+
+    private static final Map<String, String> TEXTS = Map.ofEntries(
+            entry("dynamic-channel.automation.after-sunset", "{0} after sunset"),
```
Again, seems quite artificial here. Perhaps `error.knx-unknown-ip-connection-type` would be a better choice for an example with substitution?

jlaur reviewed (changes requested):
I've added some minor comments. Let's merge after you had a chance to have a look. :-)

holgerfriedrich commented:
> I've added some minor comments. Let's merge after you had a chance to have a look. :-)

@jlaur, thanks. Finally, you convinced me to exchanged the strings for the unit test :-)

jlaur reviewed (approved):
LGTM, thanks!

holgerfriedrich commented:
@cweitkamp maybe you could approve the translations for the new strings introduced in this PR once again? I have uploaded German translations to Crowdin. Thanks!

cweitkamp commented:
> @cweitkamp maybe you could approve the translations for the new strings introduced in this PR once again?

Done.
